Product creation AI: Show confirm alert when trying to dismiss by dragging

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10799 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR intercepts the swipe-to-dismiss gesture on the product creation AI sheet. A confirm action sheet is displayed if there are any changes to product name or description.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to a WPCom or self-hosted store with Jetpack AI installed.
- Switch to the Products tab and select "+" to add a new product.
- Select add product with AI.
- Follow the steps to add product name, features and generate details.
- Try swiping-to-dismiss the sheet, an action sheet should be displayed asking to confirm the dismissal.
- Navigate back to previous screen, the same sheet should be displayed when attempting to swipe to dismiss.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/fe66f6c3-08f3-43f6-baba-130f02473200



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add new property to check if there's any change to the product creation AI form

itsmeichigo committed:
Intercept swipe to dismiss gesture on the product creation flow

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Container/AddProductWithAIContainerViewModel.swift:
```diff
@@ -23,6 +23,10 @@ final class AddProductWithAIContainerViewModel: ObservableObject {
     let source: AddProductCoordinator.Source
 
+    var canBeDismissed: Bool {
+        productName.isEmpty && productFeatures.isEmpty && productDescription == nil
```
Nit: Consider adding unit tests for this.

selanthiraiyan reviewed (commented):
I think there exists an edge case in which navigating back after deleting the features and name text field still shows the confirmation alert. 

i.e. The name and features fields are empty, but the confirmation alter is displayed. 

https://github.com/woocommerce/woocommerce-ios/assets/524475/13074d5c-97b7-4338-8cea-cb33729af133

itsmeichigo committed:
Update AddProductNameWithAIViewModel to not include callback closures

itsmeichigo committed:
Remove deprecated tests for AddProductNameWithAIViewModelT

itsmeichigo committed:
Add tests for canBeDismissed

itsmeichigo committed:
Update product name field with generated product name

itsmeichigo committed:
Merge branch 'trunk' into feat/10799-confirm-alert-on-dimissal

itsmeichigo committed:
Fix test failures

itsmeichigo commented:
Thanks for the review @selanthiraiyan! To check the state of the product name screen, I've moved its view model to the container view's view model. I updated the check and logic of both screens in the subsequent commits - this PR is ready for another look.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift:
```diff
@@ -2,6 +2,45 @@ import XCTest
+        XCTAssertTrue(viewModel.canBeDismissed)
+    }
+
+    func test_canBeDismissed_returns_False_if_current_step_is_productName_and_the_name_field_is_not_empty() {
```
Super nit:
```suggestion
    func test_canBeDismissed_returns_false_if_current_step_is_productName_and_the_name_field_is_not_empty() {
```

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/AddProductWithAIContainerViewModelTests.swift:
```diff
@@ -2,6 +2,45 @@ import XCTest
+        XCTAssertFalse(viewModel.canBeDismissed)
+    }
+
+    func test_canBeDismissed_returns_False_if_current_step_is_not_product_name() {
```
Super nit:
```suggestion
    func test_canBeDismissed_returns_false_if_current_step_is_not_product_name() {
```

selanthiraiyan reviewed (approved):
Thank you for the quick updates. LGTM. 🚀

itsmeichigo committed:
Fix typo for unit tests

itsmeichigo committed:
Merge branch 'trunk' into feat/10799-confirm-alert-on-dimissal

itsmeichigo committed:
Remove redundant parameters
